### PR TITLE
Ada Bindings for wolfSSL

### PR DESCRIPTION
**Pull request by dgarske**
# Description

Ada Bindings for wolfSSL. TLS server and client.

Credit @joakim-strandberg Joakim Dahlgren Strandberg <joakimds@kth.se>

```
Summary of SPARK analysis
=========================

---------------------------------------------------------------------------------------------------------------
SPARK Analysis results        Total        Flow   CodePeer                       Provers   Justified   Unproved
---------------------------------------------------------------------------------------------------------------
Data Dependencies                 2           2          .                             .           .          .
Flow Dependencies                 .           .          .                             .           .          .
Initialization                   15          15          .                             .           .          .
Non-Aliasing                      .           .          .                             .           .          .
Run-time Checks                  52           .          .    52 (CVC4 85%, Trivial 15%)           .          .
Assertions                        2           .          .                      2 (CVC4)           .          .
Functional Contracts             88           .          .                     88 (CVC4)           .          .
LSP Verification                  .           .          .                             .           .          .
Termination                       .           .          .                             .           .          .
Concurrency                       .           .          .                             .           .          .
---------------------------------------------------------------------------------------------------------------
Total                           159    17 (11%)          .                     142 (89%)           .          .


max steps used for successful proof: 1
```

# Testing

See wrapper/Ada/README.md

# Checklist

 - [ ] added tests
 - [ ] updated/added doxygen
 - [ ] updated appropriate READMEs
 - [ ] Updated manual and documentation

### CODE REVIEW

**Comment by dgarske** [comment]
FYI: Only issue is C++ style comments. Please use C style `/* */`. Thanks

```
[check-source-text] [2 of 2] [8ab723e42e]
C++-style comments:
./wrapper/Ada/c_tls_server_main.c:147:// To execute the application: ./main
./wrapper/Ada/ada_binding.c:27:// These functions give access to the integer values of the enumeration
./wrapper/Ada/ada_binding.c:28:// constants used in WolfSSL. These functions make it possible
./wrapper/Ada/ada_binding.c:29:// for the WolfSSL implementation to change the values of the constants
./wrapper/Ada/ada_binding.c:30:// without the need to make a corresponding change in the Ada code.
    check-source-text fail_C
```

**Comment by joakim-strandberg** [comment]
I have added the Ada tls v1.3 client application and updated the WolfSSL Ada package. I seem to have trouble uploading the changes:
joakim@joakim-N8xEJEK:~/wolfssl/wrapper/Ada$ git commit -m 'Ada version of TLS v1.3 client application implemented'
[ada 45d8a5b04] Ada version of TLS v1.3 client application implemented
 12 files changed, 657 insertions(+), 94 deletions(-)
 create mode 100644 wrapper/Ada/client.gpr
joakim@joakim-N8xEJEK:~/wolfssl/wrapper/Ada$ git status
På grenen ada
inget att checka in, arbetskatalogen ren
joakim@joakim-N8xEJEK:~/wolfssl/wrapper/Ada$ git push origin ada
Username for 'https://github.com': joakim-strandberg
Password for 'https://joakim-strandberg@github.com': 
remote: Permission to wolfSSL/wolfssl.git denied to joakim-strandberg.
fatal: kan inte komma åt "https://github.com/wolfSSL/wolfssl/": The requested URL returned error: 403
joakim@joakim-N8xEJEK:~/wolfssl/wrapper/Ada$

To get the source code locally on my computer I followed the instructions at: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/checking-out-pull-requests-locally

git clone https://github.com/wolfSSL/wolfssl
cd wolfssl/
git fetch origin pull/6605/head:ada
git checkout ada

Will be away from keyboard for a couple of hours and will try again later.

File: wrapper/Ada/README.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+./gnat-2021-20210519-x86_64-linux-bin
+```
+
+```sh
+export PATH="/opt/GNAT/2021/bin:$PATH"
+gprclean
+gprbuild default.gpr
+
+
+./c_tls_server_main
```
Comment (dgarske) [inline wrapper/Ada/README.md:19]: @joakim-strandberg can you expand this section and update with latest steps? Perhaps also add some information about Ada and advantages for using?

Comment (joakim-strandberg) [reply wrapper/Ada/README.md:19]: Sure!

File: wrapper/Ada/README.md [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+Linux Install:
+
+```sh
+chmod +x gnat-2021-20210519-x86_64-linux-bin
+./gnat-2021-20210519-x86_64-linux-bin
+```
+
+```sh
+export PATH="/opt/GNAT/2021/bin:$PATH"
```
Comment (JacobBarthelmeh) [inline wrapper/Ada/README.md:65]: I'd like the readme build instructions to be as close to copy and paste the commands into a terminal and it does a default build.

Can we add that gprclean is ran from the wrapper/Ada path. `cd wrapper/Ada`.

File: wrapper/Ada/README.md [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+gprclean
+gprbuild default.gpr
```
Comment (JacobBarthelmeh) [inline wrapper/Ada/README.md:67]: This fails for me:


```
   [Ada]          tls_client.adb
tls_client.adb:157:40: error: violation of restriction "No_Secondary_Stack" at /home/jak/Documents/wolfssl/wrapper/Ada/gnat.adc:1
gprbuild: *** compilation phase failed

```

I see the secondary stack mentioned earlier in the readme, do I need to make an adjustment to a config file for it to build?

File: wrapper/Ada/README.md [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+gprbuild client.gpr
```
Comment (JacobBarthelmeh) [inline wrapper/Ada/README.md:68]: The command fails for me with :

```
   [C]            quic.c
   [C]            t.c
/home/jak/Documents/wolfssl/t.c:6:5: warning: no previous prototype for 'test_1' [-Wmissing-prototypes]
    6 | int test_1(void)
      |     ^~~~~~
/home/jak/Documents/wolfssl/t.c:11:6: warning: no previous prototype for 'test_2' [-Wmissing-prototypes]
   11 | void test_2(int a, int b)
      |      ^~~~~~
/home/jak/Documents/wolfssl/t.c: In function 'test_2':
/home/jak/Documents/wolfssl/t.c:11:24: warning: unused parameter 'b' [-Wunused-parameter]
   11 | void test_2(int a, int b)
      |                    ~~~~^
/home/jak/Documents/wolfssl/t.c: At top level:
/home/jak/Documents/wolfssl/t.c:16:5: warning: function declaration isn't a prototype [-Wstrict-prototypes]
   16 | int main()
      |     ^~~~
/home/jak/Documents/wolfssl/t.c: In function 'main':
/home/jak/Documents/wolfssl/t.c:4:19: error: expected ')' before ';' token
    4 | #define Y test_1(); X
      |                   ^
/home/jak/Documents/wolfssl/t.c:2:28: note: in definition of macro 'test'
    2 | #define test(a, b) test_2((a), b)
      |                            ^
/home/jak/Documents/wolfssl/t.c:18:10: note: in expansion of macro 'Y'
   18 |     test(Y, X);
      |          ^
/home/jak/Documents/wolfssl/t.c:2:27: note: to match this '('
    2 | #define test(a, b) test_2((a), b)
      |                           ^
/home/jak/Documents/wolfssl/t.c:18:5: note: in expansion of macro 'test'
   18 |     test(Y, X);
      |     ^~~~
gprbuild: *** compilation phase failed

```

Am I missing a step?

**Review by JacobBarthelmeh — CHANGES_REQUESTED** [review]

File: wrapper/Ada/README.md (hunk 2 above)
Comment (joakim-strandberg) [reply wrapper/Ada/README.md:65]: Thanks for the feedback, I will update.

File: wrapper/Ada/README.md (hunk 3 above)
Comment (joakim-strandberg) [reply wrapper/Ada/README.md:67]: I will simplify this. Thanks!

File: wrapper/Ada/README.md (hunk 4 above)
Comment (JacobBarthelmeh) [reply wrapper/Ada/README.md:69]: This is from me not starting with a clean directory when trying to build, it picked up other source files.

**Review by JacobBarthelmeh — APPROVED** [review]

File: wrapper/Ada/README.md (hunk 4 above)
Comment (dgarske) [reply wrapper/Ada/README.md:69]: That happened to me too. It might be good to put a note in the readme.

**Comment by JacobBarthelmeh** [comment]
Thanks @joakim-strandberg works good. Not seeing the contributor documents in our internal tracker that we need in place yet though to merge. I'll be following up on that internally.

**Pull request merged**